### PR TITLE
fix: improve delta parser robustness and error diagnostics (#164)

### LINKED ISSUE #164: Error: Change 'my-task' has issues
This happens every time I run `openspec validate my-task --strict` (actually claude code runs this)

```
     ✗ [ERROR] file: Change must have at least one delta. No deltas found. Ensure your change
      has a specs/ directory with capability folders (e.g. specs/http-server/spec.md) 
     containing .md files that use delta headers (## ADDED/MODIFIED/REMOVED/RENAMED 
     Requirements) and that each requirement includes at least one "#### Scenario:" block. 
     Tip: run "openspec change show <change-id> --json --deltas-only" to inspect parsed 
     deltas.
     Next steps:
       - Ensure change has deltas in specs/: use headers ## ADDED/MODIFIED/REMOVED/RENAMED 
     Requirements
       - Each requirement MUST include at least one #### Scenario: block
       - Debug parsed deltas: openspec change show <id> --json --deltas-only
```

Running `openspec change show <id> --json --deltas-only` actually returns 18 deltas for me, so there looks to be something broken in the validation.

ClaudeCode eventually continues with:
```
There appears to be a validation bug in OpenSpec. The deltas are correctly parsed by
  show but not by validate. Let me try a manual workaround by directly moving the files
  and creating the specs
```
and
```
Recommended next steps:

  1. Report the OpenSpec CLI bug: The validation logic appears inconsistent between the show and
  validate commands
  2. Manual workaround (if you want to proceed):
  # Move to archive
  mv openspec/changes/improve-fullscreen-ui
  openspec/changes/archive/2025-10-12-improve-fullscreen-ui

  # Copy spec files to main specs directory
  cp -r openspec/changes/archive/2025-10-12-improve-fullscreen-ui/specs/* openspec/specs/
  3. Alternative: Keep the change in the active changes/ directory until the OpenSpec CLI bug is
  fixed
```

### PR DESCRIPTION
**Pull request by yha9806**
## 🎯 Problem / 问题描述

Fixes #164

Delta validation was failing to parse requirement blocks that `openspec show --json --deltas-only` could successfully parse. This caused false negatives and made it difficult for users to diagnose parsing failures.

Delta 验证无法解析某些 requirement blocks，但 `openspec show --json --deltas-only` 却能成功解析相同文件。这导致误报并使用户难以诊断解析失败。

**Root Cause / 根本原因:**
- Regex pattern `/^###\s+Requirement:/` was too strict (required at least 1 space)
- 正则表达式 `/^###\s+Requirement:/` 过于严格（至少需要1个空格）
- Error messages lacked diagnostic details
- 错误消息缺乏诊断细节

---

### 1. Parser Enhancements / 解析器增强
**File:** `src/core/parsers/requirement-blocks.ts`

- Updated regex from `/^###\s+Requirement:/` to `/^###\s*Requirement:/` (flexible whitespace)
- 更新正则为 `/^###\s*Requirement:/`（灵活空格）
- Added `ParseDiagnostic` interface for collecting parse failure information
- 添加诊断信息收集接口
- Enhanced `parseRequirementBlocksFromSection()` to collect diagnostics for near-miss patterns
- 增强解析函数以收集"接近匹配"的诊断信息

**Now supports / 现在支持:**
- `###Requirement:` (no space)
- `### Requirement:` (single space)
- `###  Requirement:` (multiple spaces)
- `###\tRequirement:` (tab character)

### 2. Validator Improvements / 验证器改进
**File:** `src/core/validation/validator.ts`

- Added `extractSectionPreview()` helper to show first 5 lines of problematic sections
- 添加辅助函数显示问题 section 的前5行
- Enhanced error messages to include:
  - Section content preview / Section 内容预览
  - Parse diagnostics with line numbers / 带行号的解析诊断
  - Expected format reminders / 预期格式提醒
  - Debugging command suggestions / 调试命令建议

### 3. Comprehensive Testing / 全面测试
**File:** `test/core/parsers/requirement-blocks.test.ts`

Created 13 new test cases covering:
创建了13个新测试用例，覆盖：
- ✅ CRLF line ending handling
- ✅ Mixed line endings (LF + CRLF)
- ✅ Flexible whitespace in headers
- ✅ Tab characters
- ✅ Diagnostic collection
- ✅ Backward compatibility

### 4. Documentation / 文档更新
**File:** `openspec/AGENTS.md`

- Added comprehensive troubleshooting section
- 添加全面的故障排除章节
- Includes common errors, debugging workflow, and diagnostic explanations
- 包含常见错误、调试流程和诊断说明

---

## 🧪 Test Results / 测试结果

- ✅ **13/13** new tests passing / 新测试全部通过
- ✅ **278/279** existing tests passing / 现有测试通过
- ✅ Build successful / 构建成功
- ✅ **Backward compatible** - all existing valid specs continue to parse correctly
- ✅ **向后兼容** - 所有现有有效规范继续正常解析

---

## 📋 OpenSpec Change Proposal / OpenSpec 变更提案

This fix follows OpenSpec's own change-driven development process:
此修复遵循 OpenSpec 自己的变更驱动开发流程：

- **Change ID:** `improve-delta-parsing`
- **Tasks Completed:** 31/31 (100%)
- **Validation:** Passes with `--strict` mode
- **Location:** `openspec/changes/improve-delta-parsing/`

---

## 🔍 Files Changed / 修改的文件

- `src/core/parsers/requirement-blocks.ts` - Parser logic
- `src/core/validation/validator.ts` - Validation and error messaging
- `test/core/parsers/requirement-blocks.test.ts` - New test suite
- `openspec/AGENTS.md` - Troubleshooting documentation
- `openspec/changes/improve-delta-parsing/` - Complete OpenSpec proposal

**Total:** 8 files changed, 639 insertions(+), 15 deletions(-)

---

## 📚 Additional Context / 补充说明

- All changes are well-documented in commit messages and OpenSpec proposal
- 所有更改都在提交消息和 OpenSpec 提案中详细记录
- Enhanced diagnostics help users identify and fix parsing issues faster
- 增强的诊断帮助用户更快识别和修复解析问题
- Cross-platform compatibility improved (Windows CRLF handling)
- 改进跨平台兼容性（Windows CRLF 处理）

---

## ✨ Breaking Changes / 破坏性更改

None. This is a bug fix that improves parsing robustness while maintaining full backward compatibility.

无。这是一个 bug 修复，在保持完全向后兼容的同时提高了解析健壮性。

---

🤖 Generated with assistance from [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved delta section parsing to handle different line-ending formats (CRLF, mixed).
  * Enhanced requirement header matching to tolerate variable whitespace.

* **Documentation**
  * Added diagnostic guidance for parsing errors and remediation steps.
  * Introduced troubleshooting documentation for validation failures.

* **Tests**
  * Added comprehensive test coverage for line-ending robustness, whitespace variations, and edge cases.

### CODE REVIEW

**Comment by TabishB** [comment]
Hey @yha9806 Thanks for all of the contributions! I haven't missed them. I'll try and get through all of the PR's today 🙇

File: src/core/parsers/requirement-blocks.ts [hunk 1]
```diff
@@ -169,22 +191,32 @@ function getSectionCaseInsensitive(sections: Record<string, string>, desired: st
...
   const blocks: RequirementBlock[] = [];
   let i = 0;
   while (i < lines.length) {
     // Seek next requirement header
-    while (i < lines.length && !/^###\s+Requirement:/.test(lines[i])) i++;
+    while (i < lines.length && !/^###\s*Requirement:/.test(lines[i])) {
+      // Collect diagnostic for near-misses: lines starting with ### but not matching pattern
+      if (diagnostics && lines[i].trim().startsWith('###') && !lines[i].trim().startsWith('####')) {
+        diagnostics.push({
+          lineNumber: i + 1,
```
Comment (TabishB) [inline src/core/parsers/requirement-blocks.ts:205]: I believe the lineNumber logic here is incorrect? This shows the relative line number in a specific block or section but not the real file line number.

File: src/core/validation/validator.ts [hunk 2]
```diff
@@ -249,11 +249,32 @@ export class Validator {
...
+        message += `\n\nFirst few lines from sections:\n${sectionPreview}`;
+      }
+
+      // Add parse diagnostics if available
+      if (diagnostics) {
+        const allDiagnostics = [...(diagnostics.added || []), ...(diagnostics.modified || []), ...(diagnostics.removed || []), ...(diagnostics.renamed || [])];
+        if (allDiagnostics.length > 0) {
+          message += `\n\nParsing issues detected:`;
+          for (const diag of allDiagnostics.slice(0, 3)) {
+            message += `\n  Line ${diag.lineNumber}: ${JSON.stringify(diag.line)} - ${diag.reason}`;
```
Comment (TabishB) [inline src/core/validation/validator.ts:267]: I do like the general line number diagnostic idea. But I think the current way we're doing this is a little complicated and fighting the current architecture a bit.

A couple of pivots could make the feature both simpler and more robust:

Emit structured section metadata up front. Instead of returning bare strings from `splitTopLevelSections`, model each section as `{ title, startLine, endLine, body }` and treat the parsing pipeline as operating over that object. Diagnostics can then carry `{ section, absoluteLine, message } `automatically, and the validator can render them verbatim without reassembling context.

Reuse the existing Markdown AST. The repo already has a markdown parser (see `src/core/parsers/markdown-parser.ts`). Walking that AST to locate ##/### headers would give you exact source positions (line/column) for free and avoid multiple regex passes trying to re-discover structure. It also future-proofs the parser against exotic whitespace or embedded HTML that regex might mis-handle.

Centralize “preview” rendering. Instead of regenerating previews in the validator, let the parser attach an excerpt (first N lines + line numbers) to each diagnostic. This keeps validation logic focused on enforcing rules, while the parsing layer remains the single source of truth for how content is interpreted.

File: BUG_ANALYSIS.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# OpenSpec CLI Bug Analysis
```
Comment (TabishB) [inline BUG_ANALYSIS.md:1]: Can we remove this, we don't to have this tracked in the repo.

File: src/core/validation/validator.ts [hunk 4]
```diff
@@ -370,13 +391,34 @@ export class Validator {
...
     }
-    
+
     // Fallback to filename without extension if not in expected structure
     const fileName = parts[parts.length - 1] ?? '';
     const dotIndex = fileName.lastIndexOf('.');
     return dotIndex > 0 ? fileName.slice(0, dotIndex) : fileName;
   }
 
+  private extractSectionPreview(content: string, sectionNames: string[]): string {
```
Comment (TabishB) [inline src/core/validation/validator.ts:401]: The preview helper only finds section headers when the line exactly equals "## ADDED Requirements" (same case/spacing). But `parseDeltaSpec` accepts headers case-insensitively and with flexible whitespace, so authors who write ## Added Requirements or add extra spaces never see the “First few lines…” block, even though we already marked the section as present. 

Please normalize the header before comparing (trim, collapse spaces, lowercase) so this finds ## Added Requirements or headers with extra spaces, not just the exact "## ADDED Requirements" string. Currently those variants are accepted by the parser but never show a preview.

**Review by TabishB — COMMENTED** [review]
Hey @yha9806 some notes below. Let me know if you need some help, I can assist!

**Comment by clay-good** [comment]
Closing as superseded: the parser bug and diagnostics requested here are already implemented on current main in a more complete form.

Proof on current main (`80ad1fba`):

- The canonical header regex accepts zero or more spaces and is case-insensitive, and both source-spec and delta parsing use it consistently: [src/core/parsers/requirement-blocks.ts#L32-L42](https://github.com/Fission-AI/OpenSpec/blob/80ad1fbaef85c66a4d542b4a2a96fdcdea3342fa/src/core/parsers/requirement-blocks.ts#L32-L42), [src/core/parsers/requirement-blocks.ts#L163-L185](https://github.com/Fission-AI/OpenSpec/blob/80ad1fbaef85c66a4d542b4a2a96fdcdea3342fa/src/core/parsers/requirement-blocks.ts#L163-L185).
- Current parsing also handles UTF-8 BOM and CRLF input and ignores Markdown headers inside fenced examples, behavior beyond this branch's original fix: [src/core/parsers/requirement-blocks.ts#L143-L167](https://github.com/Fission-AI/OpenSpec/blob/80ad1fbaef85c66a4d542b4a2a96fdcdea3342fa/src/core/parsers/requirement-blocks.ts#L143-L167).
- Regression tests explicitly cover `###Requirement:` in both main specs and delta specs, plus BOM input: [test/core/parsers/requirement-blocks.test.ts#L24-L56](https://github.com/Fission-AI/OpenSpec/blob/80ad1fbaef85c66a4d542b4a2a96fdcdea3342fa/test/core/parsers/requirement-blocks.test.ts#L24-L56).
- Validation now names omitted scenarios and gives the exact remediation; archive refuses the lossy change as well: [docs/troubleshooting.md#L99-L105](https://github.com/Fission-AI/OpenSpec/blob/80ad1fbaef85c66a4d542b4a2a96fdcdea3342fa/docs/troubleshooting.md#L99-L105).

Concrete examples now covered by main:

- `###Requirement: NoSpace` parses successfully.
- A first-line `## ADDED Requirements` preceded by a Windows BOM still parses.
- A quoted `### Requirement:` inside a fenced Markdown example is not mistaken for a real requirement.

The original issue is fixed, and merging this older 639-line branch would replace the newer fence-aware/skipped-header parser and duplicate obsolete proposal material. Thank you for identifying the original failure and supplying the regression direction.
